packages/flutter/lib/src/cupertino/segmented_control.dart

```diff
@@ -0,0 +1,547 @@
+        super(key: key);
+
+  /// The identifying keys and corresponding widget values in the
+  /// segmented control.
+  ///
+  /// The map must have more than one entry.
+  /// This attribute must be an ordered [Map] such as a [LinkedHashMap].
+  final Map<T, Widget> children;
+
+  /// The identifier of the widget that is currently selected.
+  ///
+  /// This must be one of the keys in the [Map] of [children].
+  /// If this attribute is null, no widget will be initially selected.
+  final T groupValue;
+
+  /// The callback that is called when a new option is tapped.
+  ///
+  /// This attribute must not be null.
+  ///
+  /// The segmented control passes the newly selected widget's associated key
+  /// to the callback but does not actually change state until the parent
+  /// widget rebuilds the segmented control with the new [groupValue].
+  ///
+  /// The callback provided to [onValueChanged] should update the state of
+  /// the parent [StatefulWidget] using the [State.setState] method, so that
+  /// the parent gets rebuilt; for example:
+  ///
+  /// ## Sample code
+  ///
+  /// ```dart
+  /// class SegmentedControlExample extends StatefulWidget {
+  ///   @override
+  ///   State createState() => new SegmentedControlExampleState();
+  /// }
+  ///
+  /// class SegmentedControlExampleState extends State<SegmentedControlExample> {
+  ///   final Map<int, Widget> children = const {
+  ///     0: const Text('Child 1'),
+  ///     1: const Text('Child 2'),
+  ///   };
+  ///
+  ///   int currentValue;
+  ///
+  ///   @override
+  ///   Widget build(BuildContext context) {
+  ///     return new Container(
+  ///       child: new SegmentedControl<int>(
+  ///         children: children,
+  ///         onValueChanged: (int newValue) {
+  ///           setState(() {
+  ///             currentValue = newValue;
+  ///           });
+  ///         },
+  ///         groupValue: currentValue,
+  ///       ),
+  ///     );
+  ///   }
+  /// }
+  /// ```
+  final ValueChanged<T> onValueChanged;
+
+  @override
+  _SegmentedControlState<T> createState() => _SegmentedControlState<T>();
+}
+
+class _SegmentedControlState<T> extends State<SegmentedControl<T>> {
+  int _selectedIndex;
+  int _pressedIndex;
+  T _selectedKey;
+  T _pressedKey;
+
+  @override
+  void initState() {
+    super.initState();
+    _selectedKey = widget.groupValue;
+  }
+
+  void _onTapDown(T currentKey) {
+    setState(() {
+      _pressedKey = currentKey;
+    });
+  }
+
+  void _onTapUp(TapUpDetails event) {
+    setState(() {
+      _pressedKey = null;
+    });
+  }
+
+  void _onTapCancel() {
+    setState(() {
+      _pressedKey = null;
+    });
+  }
+
+  void _onTap(T currentKey) {
+    setState(() {
+      if (currentKey != widget.groupValue) {
+        widget.onValueChanged(currentKey);
+      }
+    });
+  }
+
+  @override
+  void didUpdateWidget(SegmentedControl<T> oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    _selectedKey = widget.groupValue;
```

Comment:
if you just replace all mentions of `_selectedKey` with `widget.groupValue` i think you can now entirely drop the initState and didUpdateWidget methods